File: chapter-11-dynamic-sql/code-java/src/main/java/com/example/demo/repository/DynamicSqlRepository.java
package com.example.demo.repository;

import com.example.jooq.tables.pojos.Book;
import lombok.RequiredArgsConstructor;
import org.jooq.Condition;
import org.jooq.DSLContext;
import org.jooq.impl.DSL;
import org.springframework.stereotype.Repository;

import java.util.ArrayList;
import java.util.List;

import static com.example.jooq.Tables.BOOK;

/**
 * 【Chapter 11】 동적 SQL의 마법
 * 런타임에 입력된 파라미터만 판단하여 WHERE 조건(Condition)을 조립합니다.
 */
@Repository
@RequiredArgsConstructor
public class DynamicSqlRepository {

    private final DSLContext dsl;

    /**
     * 패턴 1: Condition 리스트 조립
     * null이 아닌 파라미터만 리스트에 추가하고, 마지막에 DSL.and()로 묶습니다.
     * 빈 리스트면 DSL.and()는 항상 참(TRUE)이 되어 전체를 조회합니다.
     */
    public List<Book> searchBooks(String title, Integer authorId, Integer year) {
        List<Condition> conditions = new ArrayList<>();

        if (title != null && !title.isBlank()) {
            conditions.add(BOOK.TITLE.containsIgnoreCase(title)); // ILIKE '%title%'
        }
        if (authorId != null) {
            conditions.add(BOOK.AUTHOR_ID.eq(authorId));
        }
        if (year != null) {
            conditions.add(BOOK.PUBLISHED_YEAR.eq(year));
        }

        return dsl.selectFrom(BOOK)
                  .where(DSL.and(conditions))
                  .orderBy(BOOK.TITLE.asc())
                  .fetchInto(Book.class);
    }

    /**
     * 패턴 2: DSL.noCondition() 체이닝
     * 항상 TRUE인 빈 조건에서 시작하여 .and()를 체이닝합니다.
     */
    public List<Book> searchBooksWithNoCondition(String title, Integer year) {
        Condition condition = DSL.noCondition(); // 시작점: 1 = 1 (TRUE)

        if (title != null && !title.isBlank()) {
            condition = condition.and(BOOK.TITLE.containsIgnoreCase(title));
        }
        if (year != null) {
            condition = condition.and(BOOK.PUBLISHED_YEAR.eq(year));
        }

        return dsl.selectFrom(BOOK)
                  .where(condition)
                  .orderBy(BOOK.TITLE.asc())
                  .fetchInto(Book.class);
    }
}

File: chapter-11-dynamic-sql/code-kotlin/src/main/kotlin/com/example/demo/repository/DynamicSqlRepository.kt
package com.example.demo.repository

import com.example.jooq.tables.pojos.Book
import com.example.jooq.tables.references.BOOK
import org.jooq.Condition
import org.jooq.DSLContext
import org.jooq.impl.DSL
import org.springframework.stereotype.Repository

/**
 * 【Chapter 11】 동적 SQL의 마법 (Kotlin)
 * Kotlin의 takeIf, let, fold를 활용하여 더욱 우아하게 동적 조건을 조립합니다.
 */
@Repository
class DynamicSqlRepository(
    private val dsl: DSLContext
) {

    /**
     * 패턴 1: mutableListOf + takeIf / let 조합
     * null이 아닌 파라미터만 판단하여 조건 리스트에 추가합니다.
     */
    fun searchBooks(title: String?, authorId: Int?, year: Int?): List<Book> {
        val conditions = mutableListOf<Condition>()

        title?.takeIf { it.isNotBlank() }
             ?.let { conditions.add(BOOK.TITLE.containsIgnoreCase(it)) }

        authorId?.let { conditions.add(BOOK.AUTHOR_ID.eq(it)) }

        year?.let { conditions.add(BOOK.PUBLISHED_YEAR.eq(it)) }

        return dsl.selectFrom(BOOK)
            .where(DSL.and(conditions)) // 빈 리스트인 경우 WHERE TRUE
            .orderBy(BOOK.TITLE.asc())
            .fetchInto(Book::class.java)
    }

    /**
     * 패턴 2: listOfNotNull + fold + noCondition() 조합
     * 가장 Kotlin다운 방식으로, 불변(immutable) 리스트를 조립 후 fold 연산으로 합칩니다.
     */
    fun searchBooksWithNoCondition(title: String?, year: Int?): List<Book> {
        val conditions = listOfNotNull(
            title?.takeIf { it.isNotBlank() }?.let { BOOK.TITLE.containsIgnoreCase(it) },
            year?.let { BOOK.PUBLISHED_YEAR.eq(it) }
        )

        // 초기값 noCondition()에 모든 Condition을 and로 누적 결합
        val where = conditions.fold(DSL.noCondition()) { acc, cond -> acc.and(cond) }

        return dsl.selectFrom(BOOK)
            .where(where)
            .orderBy(BOOK.TITLE.asc())
            .fetchInto(Book::class.java)
    }
}
